### ISSUE TITLE
Add batch greedy training workflow for QoI models with error estimates

### ISSUE THREAD

**Issue by eparish1** [Pressio/rom-tools-and-workflows#326]
## Motivation

The existing `romtools.workflows.greedy` workflow performs sequential error-greedy training: it evaluates a `QoiModelWithErrorEstimate` over a candidate parameter set, selects the parameter with the largest error estimate, runs the FOM there, rebuilds the ROM, and repeats.

For expensive simulations, it is often preferable to select several new training points at once so that the corresponding FOM evaluations can be launched concurrently. We should add a **batch greedy** workflow that sits next to the current greedy workflow and uses the same QoI-model-with-error-estimate API.

The batch-selection algorithm should be agnostic to how the error estimate is produced. The model may provide a residual-based estimator, learned error model, GP variance/standard deviation, ensemble disagreement, or another indicator. The workflow should only assume that larger returned error estimates indicate more valuable training locations.

## Proposed location

Add a sibling package:

```text
romtools/workflows/
  greedy/
  batch_greedy/
    __init__.py
    run_batch_greedy.py
```

with corresponding tests under:

```text
tests/romtools/workflows/batch_greedy/
```

The implementation should reuse/refactor common machinery from `romtools.workflows.greedy` rather than duplicating model execution, bookkeeping, or error-estimate evaluation.

## Proposed algorithm

At outer iteration `k`, let the current QoI model with error estimate provide an error indicator

```text
e(mu) >= 0
```

for each candidate parameter `mu` in the current testing/candidate set.

For `batch_size = 1`, the method should reduce exactly to the existing sequential greedy rule:

```text
mu_1 = argmax_mu e(mu)
```

For `batch_size > 1`, construct the batch sequentially **without rebuilding the ROM inside the batch**. The first point is the largest-error candidate. Subsequent points balance error and diversity:

```text
mu_j = argmax_mu e(mu) * d(mu, B_{j-1})**alpha
```

where

```text
d(mu, B) = min_{z in B} ||mu_tilde - z_tilde||_2
```

and `mu_tilde` denotes normalized parameter coordinates. `B_{j-1}` contains the points already selected for the current batch.

Important: the distance penalty should initially be applied only with respect to points selected in the **current batch**, not the existing training set. The error-estimate model already has access to the training history; if it reports a large error near an existing training point, the workflow should not automatically suppress that location.

After selecting the batch:

1. run the FOM at every selected parameter;
2. compute/store the corresponding FOM QoIs;
3. add all selected run directories to the training set;
4. rebuild the QoI model once using the enlarged training set;
5. update/calibrate any QoI-error model using all newly available FOM/ROM comparisons;
6. replenish the candidate set consistently with the existing greedy workflow;
7. repeat until the stopping criterion is satisfied.

This gives one ROM rebuild per batch rather than one rebuild per new FOM sample and makes the expensive FOM evaluations naturally parallelizable.

## Proposed public API

A first implementation could mirror `run_greedy` closely:

```python
def run_batch_greedy(
    fom_model: QoiModel,
    rom_model_builder: QoiModelWithErrorEstimateBuilder,
    parameter_space: ParameterSpace,
    absolute_greedy_work_directory: str,
    tolerance: float,
    batch_size: int,
    testing_sample_size: int = 10,
    random_seed: int = 1,
    calibrated_error: bool = True,
    distance_exponent: float = 1.0,
    evaluation_concurrency: int = 1,
    dispatcher: Optional[BaseDispatcher] = None,
):
    ...
```

The exact concurrency/dispatcher arguments should follow the current workflow execution conventions rather than introducing a separate execution mechanism.

`distance_exponent` controls the strength of within-batch diversity:

- `distance_exponent = 0` gives top-`q` error selection;
- larger values increasingly favor separated points;
- `batch_size = 1` is identical to ordinary error greedy regardless of this value.

Parameter distances should be computed after normalizing each parameter dimension to a common scale. Prefer using bounds/scaling information already available through `ParameterSpace`; if that is not generally available, add a small internal normalization utility with clearly documented behavior.

## Error-estimate handling

Do **not** introduce a GP-specific acquisition interface or a second error-estimator abstraction. The workflow should use the existing QoI model with error estimate returned by `QoiModelWithErrorEstimateBuilder`, e.g.

```python
rom_model.compute_error_estimate(run_directory, parameter_dict)
```

The batch-greedy code should not know whether this quantity came from a GP, deterministic estimator, learned model, variance estimate, etc.

For vector-valued error estimates, either preserve the existing greedy behavior if one is already defined or require `compute_error_estimate` to return the scalar acquisition/error indicator expected by the current greedy workflow. Avoid embedding a new vector-reduction policy in batch greedy unless necessary.

## Stopping criterion

Use the same calibrated QoI-error stopping logic as the current greedy workflow. The convergence check should be based on the maximum estimated/calibrated error over the candidate set **before launching a new batch**.

If the workflow has not converged but fewer than `batch_size` candidates remain, select all remaining candidates rather than failing.

## Execution and persistence

Preserve the existing greedy directory and restart/statistics conventions where practical. Extend the saved statistics with enough batch metadata to reconstruct the training history, for example:

```text
batch_indices
batch_sizes
selected_sample_indices
error_indicators_at_selection
```

The on-disk layout should clearly group ROM iterations/batches while retaining FOM run directories that can be reused for rebuilding the model.

FOM evaluations within a selected batch should support the repository's existing concurrency/dispatcher mechanism so the main benefit of batch selection can be realized on HPC systems.

## Suggested implementation structure

Before adding `run_batch_greedy`, refactor reusable pieces of `run_greedy.py` where useful. In particular, candidates for shared helpers include:

- evaluate a QoI model with error estimate over a set of candidate samples;
- execute FOM samples and collect QoIs;
- build/rebuild a model from training directories;
- candidate-set bookkeeping;
- calibrated QoI-error regression;
- statistics persistence.

Keep the batch-specific selection logic small and independently testable, e.g.

```python
def select_batch(
    parameter_samples: np.ndarray,
    error_estimates: np.ndarray,
    batch_size: int,
    distance_exponent: float = 1.0,
) -> np.ndarray:
    ...
```

This helper should greedily select the first maximum-error point and then update only the diversity factor for subsequent batch members; the underlying error estimates remain fixed until the next outer iteration/model rebuild.

## Tests

Add focused unit tests for the selection rule and workflow-level tests.

### Batch-selection tests

- `batch_size=1` returns the maximum-error candidate.
- `distance_exponent=0` returns the top-`q` candidates by error.
- distance penalization prevents a batch from collapsing onto nearly coincident high-error candidates when a comparably high-error separated candidate exists.
- parameter normalization makes the result insensitive to simple changes of physical units/scales.
- ties are deterministic.
- requesting a batch larger than the remaining candidate set returns all remaining candidates.
- invalid batch sizes and distance exponents fail clearly.

### Workflow tests

- `batch_size=1` reproduces the sequential greedy selection history for a deterministic toy problem.
- all FOM points in one batch are selected from error estimates generated by the same ROM iteration.
- the ROM/model builder is called once per completed batch, not once per sample.
- all selected FOM directories are included in the subsequent training set.
- calibrated error handling incorporates every completed point in the batch.
- convergence is checked before launching the next batch.
- fixed random seeds give deterministic candidate generation and selection.
- concurrent evaluation preserves the association between parameter samples, run directories, and QoIs.

## Documentation/example

Add a compact example alongside the existing greedy-training documentation showing how changing `batch_size` allows several expensive training simulations to be launched per adaptive iteration.

It would be useful to compare `batch_size=1` and a small batch (e.g. 4) on the same toy QoI/error model and show the selected parameter locations or convergence history.

## Acceptance criteria

- [ ] Add `romtools.workflows.batch_greedy` next to `romtools.workflows.greedy`.
- [ ] Use the existing `QoiModelWithErrorEstimateBuilder` / `compute_error_estimate` API; no GP-specific dependency.
- [ ] Implement error-greedy batch selection with within-batch distance penalization.
- [ ] Normalize parameter coordinates before computing diversity distances.
- [ ] `batch_size=1` reduces to standard sequential error greedy.
- [ ] `distance_exponent=0` reduces to top-`q` error selection.
- [ ] Rebuild the ROM/QoI model once after each completed batch.
- [ ] Support concurrent execution of FOM evaluations using the repository's existing execution infrastructure.
- [ ] Reuse/refactor common greedy machinery rather than maintaining two large duplicated workflows.
- [ ] Persist batch-selection and convergence statistics.
- [ ] Add unit and workflow tests covering selection, determinism, calibration, and concurrency.
- [ ] Add API documentation and a runnable example.

## Follow-on possibilities

Keep these outside the first implementation unless they fall out naturally from the design:

- exclusion-radius batch selection;
- error-weighted clustering / k-center selection;
- alternative distance metrics or user-provided feature-space distances;
- stochastic/fantasized updates for probabilistic models such as GPs;
- adaptive batch-size selection.

The initial implementation should stay focused on a general, model-agnostic **error-greedy + diversity** workflow built on the existing QoI-model-with-error-estimate API.
